### PR TITLE
[SPARK-37923][SQL] Generate partition transforms for BucketSpec inside parser

### PR DESCRIPTION
**Pull request by huaxingao**
### What changes were proposed in this pull request?
We currently generate partition transforms for BucketSpec in Analyzer. It's cleaner to do this inside Parser.

### Why are the changes needed?
code simplification

### Does this PR introduce _any_ user-facing change?
No

### How was this patch tested?
existing tests

### CODE REVIEW

**Comment by huaxingao** [comment]
@cloud-fan

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/parser/AstBuilder.scala [hunk 1]
```diff
@@ -3468,7 +3469,8 @@ class AstBuilder extends SqlBaseBaseVisitor[AnyRef] with SQLConfHelper with Logg
         s"CREATE TEMPORARY TABLE ...$asSelect, use CREATE TEMPORARY VIEW instead", ctx)
     }
 
-    val partitioning = partitionExpressions(partTransforms, partCols, ctx)
+    val partitioning =
+      partitionExpressions(partTransforms, partCols, ctx) ++ bucketSpec.map(_.asTransform)
     val tableSpec = TableSpec(bucketSpec, properties, provider, options, location, comment,
```
Comment (cloud-fan) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/parser/AstBuilder.scala:3474]: do we still need bucket spec in `TableSpec`?

Comment (cloud-fan) [reply sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/parser/AstBuilder.scala:3474]: Today we restore partition col names from the v2 partitioning, I think we can do the same for bucket spec.

Comment (huaxingao) [reply sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/parser/AstBuilder.scala]: I removed `bucketSpec` from `TableSpec` and restored  partition col names in `CatalogV2Implicits.asPartitionColumns`. Need to change a few tests.

File: sql/core/src/test/scala/org/apache/spark/sql/execution/command/PlanResolutionSuite.scala [hunk 2]
```diff
@@ -248,37 +248,17 @@ class PlanResolutionSuite extends AnalysisTest {
     }
   }
 
-  test("create table - partitioned by transforms") {
-    val transforms = Seq(
-        "bucket(16, b)", "years(ts)", "months(ts)", "days(ts)", "hours(ts)", "foo(a, 'bar', 34)",
```
Comment (cloud-fan) [inline sql/core/src/test/scala/org/apache/spark/sql/execution/command/PlanResolutionSuite.scala:253]: shall we keep the test and check `years(ts)` etc.?

File: sql/core/src/main/scala/org/apache/spark/sql/catalyst/analysis/ResolveSessionCatalog.scala [hunk 3]
```diff
@@ -538,7 +532,6 @@ class ResolveSessionCatalog(val catalogManager: CatalogManager)
       schema = schema,
       provider = Some(provider),
       partitionColumnNames = partitioning.asPartitionColumns,
-      bucketSpec = bucketSpec,
```
Comment (cloud-fan) [inline sql/core/src/main/scala/org/apache/spark/sql/catalyst/analysis/ResolveSessionCatalog.scala:541]: where do we restore the bucket spec from the v2 partitioning?

Comment (huaxingao) [reply sql/core/src/main/scala/org/apache/spark/sql/catalyst/analysis/ResolveSessionCatalog.scala:541]: here https://github.com/apache/spark/blob/master/sql/core/src/main/scala/org/apache/spark/sql/execution/datasources/v2/V2SessionCatalog.scala#L337
seems the columns in `BucketTransform` are not added to partitionColumns.

Comment (cloud-fan) [reply sql/core/src/main/scala/org/apache/spark/sql/catalyst/analysis/ResolveSessionCatalog.scala:541]: This is the code path that we fallback to v1 command, and `V2SessionCatalog` won't be reached. I think we need to restore the bucket spec here, similar to partition cols.

File: sql/catalyst/src/main/scala/org/apache/spark/sql/connector/catalog/CatalogV2Implicits.scala [hunk 4]
```diff
@@ -49,21 +51,27 @@ private[sql] object CatalogV2Implicits {
...
-        throw QueryCompilationErrors.cannotConvertTransformsToPartitionColumnsError(nonIdTransforms)
+    def convertTransforms: (Seq[String], Option[BucketSpec]) = {
+      val identityCols = new mutable.ArrayBuffer[String]
+      var bucketSpec = Option.empty[BucketSpec]
+
+      transforms.map {
+        case IdentityTransform(FieldReference(Seq(col))) =>
+          identityCols += col
+
+        case BucketTransform(numBuckets, col, sortCol) =>
```
Comment (cloud-fan) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/connector/catalog/CatalogV2Implicits.scala:62]: we should fail if there are more than one bucket transform. I think people can trigger it like `PARTITIONED BY (bucket(4, col)) CLUSTERED BY col INTO 4 BUCKETS`?

**Review by cloud-fan — APPROVED** [review]
LGTM except for one comment

**Review by cloud-fan — APPROVED** [review]

**Comment by cloud-fan** [comment]
thanks, merging to master!

**Pull request closed without merging**